src/psyclone/domain/common/transformations/kernel_module_inline_trans.py

```diff
@@ -250,68 +296,127 @@ def _prepare_code_to_inline(code_to_inline):
+                # imported. We must therefore update its interface and
+                # potentially remove the ContainerSymbol (from which it is
+                # imported) altogether.
+                csym = existing_symbol.interface.container_symbol
+                # The import of the routine symbol may be in an outer scope.
+                ctable = csym.find_symbol_table(node)
+                remove_csym = (ctable.symbols_imported_from(csym) ==
+                               [existing_symbol])
+                existing_symbol.interface = DefaultModuleInterface()
+                if remove_csym:
+                    ctable.remove(csym)
+                container.addchild(code_to_inline.detach())
+            else:
+                # The routine symbol already exists, and we know from the
+                # validation that it's a Routine. Now check if they are
+                # exactly the same.
+                for routine in container.walk(Routine, stop_type=Routine):
+                    if routine.name == caller_name:
+                        # This TransformationError happens here and not in the
+                        # validation because it needs the symbols_to_bring_in
+                        # applied to effectively compare both versions.
+                        # This will be fixed when module-inlining versioning is
+                        # implemented.
+                        if routine != code_to_inline:
+                            raise TransformationError(
+                                f"Cannot inline subroutine '{caller_name}' "
+                                f"because another, different, subroutine with "
+                                f"the same name already exists and versioning "
+                                f"of module-inlined subroutines is not "
+                                f"implemented yet.")
 
         # We only modify the kernel call name after the equality check to
         # ensure the apply will succeed and we don't leave with an inconsistent
         # tree.
-        if node.name.lower() != name:
-            node.name = name
-
+        if callee_name != caller_name:
+            if isinstance(node, CodedKern):
+                node.name = callee_name
+            else:
+                # TODO #924 - we can't currently resolve a subroutine if its
+                # name doesn't match that in the caller (as will be the case
+                # if it's being called via an Interface in Fortran). Once
+                # we can, we will need to use the newly obtained RoutineSymbol
+                # to create a new Call here to replace node.
+                # sym = container.symbol_table.lookup(callee_name)
+                raise NotImplementedError(
+                    f"Cannot module-inline call to '{caller_name}' because its"
+                    f" name does not match that of the callee: "
+                    f"'{callee_name}'. TODO #924.")
```

Comment:
I know that we can't convert this to a TransformationError, otherwise we may return normally from an inconsistently applied transformation, but could this condition be found in the validate? If we can then this can be a TransformationError there and prevent uncatched error in user scripts.

Reply:
Oh yes - moved it to the new `get_psyir_to_inline()` method. I've kept a safety catch here to make future extension easier.